### PR TITLE
feat: make `TwoAdicFriPcs` stateless

### PR DESCRIPTION
**Pull request by chokobole**
# Description

The reason for the failing `sp1-core` unittests is due to `TwoAdicFriPcs` having internal state while being run in a multithreaded environment, causing state corruption. The following states were identified:

- `ProverDataVec` in `TwoAdicFriPcs` (Rust side)
- LDE vectors in `TwoAdicFriImpl` (C++ side)

By removing these states, the test passed, but this raises concerns about performance. OPENMP loops should ideally run on a single machine, even across different processes, to avoid potential performance issues. See https://github.com/kroma-network/tachyon/commit/0508297020cf2c8a0e6fa80af723e84e57b4870e

### CODE REVIEW

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_two_adic_fri.h [hunk 1]
```diff
@@ -79,19 +70,19 @@ tachyon_sp1_baby_bear_poseidon2_two_adic_fri_coset_lde_batch(
...
- * @brief Commits to the mixed matrix created by
- * tachyon_sp1_baby_bear_poseidon2_two_adic_fri_coset_lde_batch.
+ * @brief Commits to the lde vector.
  *
  * @param pcs A pointer to the two adic fri.
+ * @param lde_vec A pointer to lde vector.
  * @param commitment A pointer to store the commitment.
  * @param prover_data A pointer to store the field merkle tree.
- * @param prover_data_vec A pointer to the field merkle tree vector.
+ .
```
Comment (GideokKim) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_two_adic_fri.h:79]:
```suggestion
 
```

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_two_adic_fri.h [hunk 2]
```diff
@@ -79,16 +70,17 @@ tachyon_sp1_baby_bear_poseidon2_two_adic_fri_coset_lde_batch(
     tachyon_baby_bear* extended_values, tachyon_baby_bear shift);
 
 /**
- * @brief Commits to the mixed matrix created by
- * tachyon_sp1_baby_bear_poseidon2_two_adic_fri_coset_lde_batch.
+ * @brief Commits to the lde vector.
  *
  * @param pcs A pointer to the two adic fri.
+ * @param lde_vec A pointer to lde vector.
```
Comment (TomTaehoonKim) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_two_adic_fri.h:76]:
```suggestion
 * @param lde_vec A pointer to the lde vector.
```
How about like this for consistency?

**Comment by TomTaehoonKim** [comment]
@chokobole Can you remove the "Since" in the beginning of the commit message of [0c63f4f](https://github.com/kroma-network/tachyon/pull/541/commits/0c63f4f68ed11435746f7ddfa9de7dd192cd3d0c) and [af10ea9](https://github.com/kroma-network/tachyon/pull/541/commits/af10ea9e086fd152ae9e5ced0cc6eac88298bbc5)? There's also since after the comma, so I think this way is much better.

File: tachyon/c/crypto/commitments/fri/two_adic_fri_impl.h [hunk 3]
```diff
@@ -38,12 +43,30 @@ class TwoAdicFRIImpl
...
+      OpenedValues* opened_values_out, FRIProof* proof) const {
+    std::vector<std::unique_ptr<ProverData>> prover_data_by_round_tmp =
+        tachyon::base::Map(prover_data_by_round, [](const ProverData* data) {
+          return std::unique_ptr<ProverData>(const_cast<ProverData*>(data));
+        });
+    CHECK(Base::CreateOpeningProof(prover_data_by_round_tmp, points_by_round,
+                                   challenger, opened_values_out, proof));
+    for (std::unique_ptr<ProverData>& prover_data : prover_data_by_round_tmp) {
+      // NOTE(chokobole): The caller is responsible for deallocating the memory.
+      // See the comment above.
```
Comment (TomTaehoonKim) [inline tachyon/c/crypto/commitments/fri/two_adic_fri_impl.h:67]: (optional) Maybe "in Commit", might be better than "above".

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Review by GideokKim — APPROVED** [review]
LGTM

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * @brief Adds lde.
+ *
+ * @param lde_vec A pointer to the lde vector.
+ * @param lde A pointer to the values.
+ * @param rows The number of the rows.
+ * @param cols The number of the cols.
+ */
+TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_lde_vec_add(
+    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec, tachyon_baby_bear* values,
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h]: I'm assuming `tachyon_baby_bear` is a single value 🤔
```suggestion
 * @param value A pointer to the value.
 * @param rows The number of the rows.
 * @param cols The number of the cols.
 */
TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_lde_vec_add(
    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec, tachyon_baby_bear* value,

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.cc [hunk 5]
```diff
@@ -21,9 +22,9 @@ void tachyon_sp1_baby_bear_poseidon2_lde_vec_destroy(
 }
 
 void tachyon_sp1_baby_bear_poseidon2_lde_vec_add(
-    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec, tachyon_baby_bear* values,
-    size_t rows, size_t cols) {
+    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec,
+    const tachyon_baby_bear* values, size_t rows, size_t cols) {
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.cc:26]:
```suggestion
    const tachyon_baby_bear* value, size_t rows, size_t cols) {

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h [hunk 6]
```diff
@@ -39,13 +39,13 @@ TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_lde_vec_destroy(
...
- * @param lde A pointer to the values.
+ * @param lde A const pointer to the values.
  * @param rows The number of the rows.
  * @param cols The number of the cols.
  */
 TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_lde_vec_add(
-    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec, tachyon_baby_bear* values,
-    size_t rows, size_t cols);
+    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec,
+    const tachyon_baby_bear* values, size_t rows, size_t cols);
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h]: lde/values -> value
```suggestion
 * @param value A const pointer to the value.
 * @param rows The number of the rows.
 * @param cols The number of the cols.
 */
TACHYON_C_EXPORT void tachyon_sp1_baby_bear_poseidon2_lde_vec_add(
    tachyon_sp1_baby_bear_poseidon2_lde_vec* lde_vec,
    const tachyon_baby_bear* value, size_t rows, size_t cols);

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_field_merkle_tree_vec.h [hunk 7]
```diff
@@ -45,6 +49,18 @@ TACHYON_C_EXPORT void
 tachyon_sp1_baby_bear_poseidon2_field_merkle_tree_vec_destroy(
     tachyon_sp1_baby_bear_poseidon2_field_merkle_tree_vec* tree_vec);
 
+/**
+ * @brief Sets field merkle tree.
+ *
+ * @param opened_values A pointer to the field merkle tree vector.
```
Comment (ashjeong) [inline tachyon/c/zk/air/sp1/baby_bear_poseidon2_field_merkle_tree_vec.h:55]:
```suggestion
 * @param tree_vec A pointer to the field merkle tree vector.

**Review by ashjeong — CHANGES_REQUESTED** [review]
https://github.com/kroma-network/tachyon/pull/541/commits/cc7b0284b15cec607ad58132d5bc57e718769976
Add a “the” before “caller” in the commit message
https://github.com/kroma-network/tachyon/pull/541/commits/b26caea96ac1316fc2dc1c7f5c4688ebef3fcd38
Change “…constructed on creating…” to “…constructed when creating...”

File: tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h (hunk 4 above)
Comment (chokobole) [reply tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h]: No it's actually pointer to the matrix with a given rows and cols!

Comment (chokobole) [reply tachyon/c/zk/air/sp1/baby_bear_poseidon2_lde_vec.h]: I'll change it to `lde` to make it clear

**Comment by dongchangYoo** [comment]
There are typo in the commit message: "breaing change -> breaking chanege"

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by ashjeong — APPROVED** [review]
LGTM

**Pull request merged**